From: Jan Kara <jack@suse.cz>
Subject: printk: Fixup console_flush() to avoid infinite loop
References: bnc#869632
Patch-mainline: No, fixes patch patches.suse/printk-Hand-over-printing-to-console-if-printing-too.patch

When console is suspended, console_unlock() doesn't flush printk buffer to
console and thus the loop in console_flush() never terminates. Fix the problem
by properly checking for suspended console.

Signed-off-by: Jan Kara <jack@suse.cz>

diff -rupX /crypted/home/jack/.kerndiffexclude linux-3.12-SLE12/kernel/printk/printk.c linux-3.12-SLE12-printk/kernel/printk/printk.c
--- linux-3.12-SLE12/kernel/printk/printk.c	2014-03-23 06:25:59.350065286 +0100
+++ linux-3.12-SLE12-printk/kernel/printk/printk.c	2014-03-23 06:31:54.806085453 +0100
@@ -2310,7 +2310,7 @@ void console_flush(void)
 		raw_spin_lock_irqsave(&logbuf_lock, flags);
 		retry = console_seq != log_next_seq;
 		raw_spin_unlock_irqrestore(&logbuf_lock, flags);
-		if (!retry)
+		if (!retry || console_suspended)
 			break;
 		/* Cycle console_sem to wait for outstanding printing */
 		console_lock();
